truffle/src/main/java/org/jruby/truffle/core/format/format/FormatIntegerNode.java

```diff
@@ -10,67 +10,203 @@
+                if (hasFSharp && value > 0) {
+                    formatted = "0X" + formatted;
+                }
+                if (value < 0) {
+                    formatted = "..F" + Integer.toString(-value, 16).toUpperCase();
+                }
+                break;
+            case 'x':
+                formatted = Integer.toString(value, 16);
+                if (hasFSharp && value > 0) {
+                    formatted = "0x" + formatted;
+                }
+                if (value < 0) {
+                    formatted = "..f" + Integer.toString(-value, 16);
+                }
+                break;
+            case 'o':
+                formatted = Integer.toString(value, 8);
+                if (hasFSharp && value > 0) {
+                    formatted = "0" + formatted;
+                }
+                if (value < 0) {
+                    formatted = "..7" + new String(elide(ConvertBytes.intToOctalBytes(value), format, precision));
+                    ;
+                }
+                break;
+            case 'd':
+            case 'i':
+            case 'u':
+                if (value < 0) {
+                    formatted = Long.toString(-value);
+                } else {
+                    formatted = Long.toString(value);
+                }
+                break;
+            default:
+                throw new UnsupportedOperationException();
+        }
+
+        return formatStart(width, precision, formatted, value < 0);
     }
 
     @TruffleBoundary
-    @Specialization(guards = "!isRubyBignum(value)", contains = "formatCached")
-    public byte[] formatUncached(int width,
-                                 Object value) {
-        return doFormat(value, makeFormatString(width));
+    @Specialization
+    public byte[] format(int width, int precision, long value) {
+        String formatted;
+        switch (this.format) {
+            case 'X':
+                formatted = Long.toString(value, 16).toUpperCase();
+                if (hasFSharp && value > 0) {
+                    formatted = "0X" + formatted;
+                }
+                if (value < 0) {
+                    formatted = "..F" + Long.toString(-value, 16).toUpperCase();
+                }
+                break;
+            case 'x':
+                formatted = Long.toString(value, 16);
+                if (hasFSharp && value > 0) {
+                    formatted = "0x" + formatted;
+                }
+                if (value < 0) {
+                    formatted = "..f" + Long.toString(-value, 16);
+                }
+                break;
+            case 'o':
+                formatted = Long.toString(value, 8);
+                if (hasFSharp && value > 0) {
+                    formatted = "0" + formatted;
+                }
+                if (value < 0) {
+                    formatted = "..7" + new String(elide(ConvertBytes.longToOctalBytes(value), format, precision));
```

Comment:
What's this `..f` and `..7` for? I've not seen that in any output that I can think of.